src/ops/ops.go

```diff
@@ -126,24 +129,67 @@ func (o *ops) SystemctlAction(action string, args ...string) error {
+func refFilePath(refName string) string {
+	return path.Join(ostreeRepo, "refs/heads", refName)
+}
 
 func (o *ops) importOSTreeCommit(liveLogger io.Writer) (string, error) {
 	ostreeReleasePullSpec := fmt.Sprintf("ostree-unverified-registry:%s", o.installerConfig.CoreosImage)
-	out, err := o.ExecPrivilegeCommand(liveLogger, "ostree", "container", "unencapsulate", "--authfile", dockerConfigFile, "--quiet", "--repo", "/ostree/repo", ostreeReleasePullSpec)
+	out, err := o.ExecPrivilegeCommand(liveLogger, "ostree", "container", "image", "pull", "--authfile", dockerConfigFile, ostreeRepo, ostreeReleasePullSpec)
+	if err != nil {
+		return "", errors.Wrapf(err, "failed to pull rhcos payload image: %s", out)
+	}
+
+	out, err = o.ExecPrivilegeCommand(liveLogger, "ostree", "refs", "--repo", ostreeRepo)
+	if err != nil {
+		return "", errors.Wrapf(err, "failed to list refs in repo")
+	}
+	refs := strings.Split(out, "\n")
+	dockerRefs := []string{}
+	for _, ref := range refs {
+		if strings.HasPrefix(ref, "ostree/container/image/docker") {
+			dockerRefs = append(dockerRefs, ref)
+		}
+	}
+	if numRefs := len(dockerRefs); numRefs != 1 {
+		return "", fmt.Errorf("found %d matching ostree refs, expected only 1: ref list: %v", numRefs, dockerRefs)
+	}
+
+	out, err = o.ExecPrivilegeCommand(liveLogger, "ostree", "refs", "--repo", ostreeRepo, dockerRefs[0], "--create", assistedInstallOSTreeRefName)
 	if err != nil {
-		return "", errors.Wrapf(err, "failed to unencapsulate rhcos payload image: %s", out)
+		return "", errors.Wrapf(err, "failed to create ref, output: %s", out)
+	}
+
+	refFile := refFilePath(assistedInstallOSTreeRefName)
+	out, err = o.ExecPrivilegeCommand(liveLogger, "cat", refFile)
+	if err != nil {
+		return "", errors.Wrapf(err, "failed to read commit from ref file %s", refFile)
+	}
+
+	return strings.TrimSpace(out), nil
+}
+
+// Removes the node image reference if it still exists, but also ensure it isn't garbage collected just yet
+func (o *ops) removeNodeImage(liveLogger io.Writer) error {
+	if !o.FileExists(refFilePath(nodeImageOSTreeRefName)) {
+		return nil
 	}
 
-	matches := ostreeOutputRegex.FindStringSubmatch(out)
-	if matches == nil {
-		return "", fmt.Errorf("got unexpected output from unencapsulate: \"%s\"", out)
+	o.log.Info("Removing existing node-image ref")
+
+	out, err := o.ExecPrivilegeCommand(liveLogger, "ostree", "refs", "--repo", ostreeRepo, "--delete", nodeImageOSTreeRefName)
+	if err != nil {
+		return errors.Wrapf(err, "failed deleting the coreos node image ref: %s", out)
+	}
+	out, err = o.ExecPrivilegeCommand(liveLogger, "touch", "/ostree/repo/tmp/node-image")
+	if err != nil {
+		return errors.Wrapf(err, "failed to update temp node image checkout: %s", out)
 	}
 
-	return matches[1], nil
+	return nil
 }
 
-func ostreeArgs(commit string, installerArgs []string) []string {
+func ostreeArgs(ref string, installerArgs []string) []string {
 	ostreeArgs := []string{"admin", "deploy",
```

Comment:
We had a live chat on this and I think you can drop a lot of the low-level ostree commands if you switch to using `ostree container image deploy` instead

Reply:
Does this command have something similar to the `--karg-append` and `--karg-delete` that I'm using below?

I see `--karg`, but not the others.

Reply:
I don't see something in the `--help`

Reply:
`--karg` is the same as `--karg-append`, now you're right it doesn't have `--karg-delete` right now; would be easy to add but can you elaborate on why it's needed?

We include almost no default kernel arguments so I am uncertain when one would need to delete any.

Reply:
Also I went down this path for a while and it seems like it's creating the content in the stateroot, but it doesn't seem to be creating the boot entry. Is that expected? Do I still need to do something else to handle that?

Reply:
Should definitely be creating the boot entry too

Reply:
>can you elaborate on why it's needed?

Adding and deleting kargs is an API we've exposed through assisted because they were options for coreos-installer. I'm just trying to keep parity here.

Reply:
I'm (effectively) running this in my last attempt:

`ostree container image deploy --stateroot install --sysroot /sysroot --image <rhcos-content-image-from-release>`

Then the next boot fails.

I booted a live ISO and mounted the boot and sysroot partitions and I see the old boot entry in the boot partition's entry, but I see the new one in the sysroot `/boot/loader/entries` directory ... This doesn't seem right. Should I have maybe specified `/` for `--sysroot` in the ostree command?

Reply:
>Yeah, try with --sysroot=/

This seems to have helped. Doing some more testing then will update the PR.

>rpm-ostree kargs --os=install --append/--delete ... should work to edit the kargs post-deployment

Nice, thanks. I'll try this out too.